cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java

```diff
@@ -0,0 +1,869 @@
+import io.strimzi.operator.PlatformFeaturesAvailability;
+import io.strimzi.operator.cluster.model.CruiseControl;
+import io.strimzi.operator.cluster.model.InvalidResourceException;
+import io.strimzi.operator.cluster.model.NoSuchResourceException;
+import io.strimzi.operator.cluster.model.StatusDiff;
+import io.strimzi.operator.cluster.operator.resource.cruisecontrol.CruiseControlApi;
+import io.strimzi.operator.cluster.operator.resource.cruisecontrol.CruiseControlApiImpl;
+import io.strimzi.operator.cluster.operator.resource.cruisecontrol.CruiseControlUserTaskResponse;
+import io.strimzi.operator.cluster.operator.resource.cruisecontrol.CruiseControlUserTaskStatus;
+import io.strimzi.operator.cluster.operator.resource.cruisecontrol.RebalanceOptions;
+import io.strimzi.operator.cluster.operator.resource.ResourceOperatorSupplier;
+import io.strimzi.operator.common.AbstractOperator;
+import io.strimzi.operator.common.Annotations;
+import io.strimzi.operator.common.Reconciliation;
+import io.strimzi.operator.common.Util;
+import io.strimzi.operator.common.model.Labels;
+import io.strimzi.operator.common.operator.resource.AbstractWatchableResourceOperator;
+import io.strimzi.operator.common.operator.resource.CrdOperator;
+import io.strimzi.operator.common.operator.resource.StatusUtils;
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import io.vertx.core.Vertx;
+import io.vertx.core.json.JsonObject;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+
+import static io.strimzi.operator.cluster.operator.resource.cruisecontrol.CruiseControlApi.CC_REST_API_SUMMARY;
+
+/**
+ * <p>Assembly operator for a "KafkaRebalance" assembly, which interacts with Cruise Control REST API</p>
+ *
+ * <p>
+ *     This operator takes care of the {@code KafkaRebalance} custom resources that a user can create in order
+ *     to interact with Cruise Control REST API and execute a cluster rebalancing.
+ *     A state machine is used for the rebalacing flow which is reflected in the {@code status} of the custom resource.
+ *
+ *     When a new {@code KafkaRebalance} custom resource is created, the operator sends a rebalance proposal
+ *     request to the Cruise Control REST API and moves to the {@code PendingProposal} state. It stays in this state
+ *     until a the rebalance proposal is ready, polling the related status on Cruise Control, and then finally moves
+ *     to the {@code ProposalReady} state. The status of the {@code KafkaRebalance} custom resource is updated with the
+ *     computed rebalance proposal so that the user can view it and making a decision to execute it or not.
+ *     For starting the actual rebalancing on the cluster, the user annotate the custom resource with
+ *     the {@code strimzi.io/rebalance=approve} annotation, triggering the operator to send a rebalance request to the
```

Comment:
We had a discussion in December about more or less how to handle this and the annotation was mentioned here https://github.com/strimzi/strimzi-kafka-operator/issues/2202 . At the end, it is like we interact to delete pod for example and annotation seems to be the more natural way instead of a field. To be honest I think that's too late for changing this now. Why do you think the field would be better than an annotation?

Reply:
The main reason why we use the annotation for deleting a pod is that there we need it to disappear after the action. So using a field inside the `.spec` would mean that we have to delete it somehow. That does not necessarily apply here. But its fine if you thought about it and think that anno is better.

Reply:
Anyway, the operator deletes the annotation when the operation is executed.